WordPress/gutenberg
# WP Scripts: Build block.json viewModule

## What?

Add handling of `viewModule` field in block.json when using `wp-scripts` `build` or `start`.

Part of https://github.com/WordPress/gutenberg/issues/57492

## Why?

As the Modules API matures, our tooling should support it.

## How?

We add an option to clearly mark this an an experimental feature: `wp-scripts build --experimental-modules`.

To support modules with webpack, it was necessary to run a multi-compilation. We use an array of webpack configurations instead of a single webpack configuration. This is because module compilation is an option at the compilation level and cannot be set for specific entrypoints.

If the `--experimental-modules` option is found, we use an environment variable to change the `webpack.config` export to return an array `[ scriptWebpackConfig, moduleWebpackConfig ]`. Without the experimental option, the `webpack.config` export is the same except for changes that should not impact the script build at all. Consumers should be able to continue extending this config without noticing any differences. I introduced a snapshot test to compare the results, the difference is currently at 20dff00fef6. I did subsequently remove this test as it doesn't seem to have long-term value.

Notable changes to the existing webpack config: **None.**

- ~`experiments.outputModule` webpack option is set to `true`. This _enables_ module compilation, but module compilation is controled by other options (`output.module`). This should not impact existing builds.~

As mentioned above, if we find the `WP_EXPERIMENTAL_MODULES` environment variable (set by adding the `--experimental-modules` option on the CLI) then the webpack config will export an array of configs.

If we run in experimental modules mode, the `CleanWebpackPlugin` could remove assets that should be included. This is because both compilations output to the same output directory. Ideally, each compilation would have control over its output directory and not share them. That's difficult and I don't have a good solution at this time given that we do want to share other assets like stylesheets. Additionally, if we were to output to different directories, the block.json `file:…` rewriting would need to account for that.

I don't have a convincing solution for either using multiple directories or preventing assets from being clobbered at this time. Given that this is clearly flagged as experimental, I believe that's acceptable and we can improve things if a convincing solution presents itself.

Since this is experimental, I wanted avoided adding more webpack configuration files. I don't want to expose more files that could folks require and extend. We could consider [adding an `exports` field to this package](https://nodejs.org/api/packages.html#exports) to control access to certain things, but that's likely a breaking change I didn't want to undertake here.

## Testing Instructions

To test this, you'll want to check this out and link a project that depends on `@wordpress/scripts` to use this PR.

If you use `wp-scripts build` or `wp-scripts start`, there should be no significant changes with the current behavior.

You can test the experimental build by adding a `viewModule` field to block.json in a project, then running a script like `wp-scripts -- build --experimental-modules` to build it.

I've been running the experimental build of https://github.com/sirreal/wp-scripts-interactivity-module-demo (the packages is linked to my development copy locally). The build produces the expected results with both scripts and modules. With #57437 (handling `viewModule` in block registration), this results in a working plugin. Here's the build ouptut:

<details>

<summary><code>block.json</code></summary>

```json
{
  "$schema": "https://schemas.wp.org/trunk/block.json",
  "apiVersion": 3,
  "title": "Jon's cool counter",
  "name": "jon/the-block",
  "editorScript": "file:./index.js",
  "viewModule": "file:./view.js",
  "render": "file:./render.php",
  "category": "text"
}
```

</details>

<details>

<summary><code>index.asset.php</code></summary>

```php
<?php return array('dependencies' => array('react', 'wp-blocks'), 'version' => '04a7cb7df938528c50ee');
```

</details>


<details>

<summary><code>index.js</code></summary>

```js
const e=window.React,t=window.wp.blocks,o=JSON.parse('{"$schema":"https://schemas.wp.org/trunk/block.json","apiVersion":3,"title":"Jon\'s cool counter","name":"jon/the-block","editorScript":"file:./index.tsx","viewModule":"file:./view.mts","render":"file:./render.php","category":"text"}');t.registerBlockType(o,{edit:()=>(0,e.createElement)("div",null,"counter"),save:()=>(0,e.createElement)("div",null,"counter")});
```

</details>


<details>

<summary><code>render.php</code></summary>

```php
<div data-wp-interactive='{"namespace":"jon/the-block"}' data-wp-context='{"val":0}'>
	<input readonly type='number' data-wp-bind--value='context.val' value='0' />
	<button data-wp-on--click='actions.inc'>+</button>
	<button data-wp-on--click='actions.dec'>-</button>
</div>
```

</details>


<details>

<summary><code>view.asset.php</code></summary>

```php
<?php return array('dependencies' => array(array('id' => '@wordpress/interactivity', 'type' => 'dynamic')), 'version' => 'c9463319dadd37e31637', 'type' => 'module');
```

</details>


<details>

<summary><code>view.js</code></summary>

```js
var e={990:e=>{e.exports=import("@wordpress/interactivity")}},t={};function n(o){var r=t[o];if(void 0!==r)return r.exports;var i=t[o]={exports:{}};return e[o](i,i.exports,n),i.exports}(()=>{async function e(){if(e.init)return;e.init=!0;const t=await Promise.resolve().then(n.bind(n,990));t.store("jon/the-block",{actions:{inc(){t.getContext().val+=1},dec(){t.getContext().val-=1}}})}document.querySelector('[data-wp-interactive*=\'"namespace":"jon/the-block"\']').addEventListener("mouseenter",e,{once:!0}),document.querySelector('[data-wp-interactive*=\'"namespace":"jon/the-block"\']').addEventListener("focusin",e,{once:!0})})();
```

</details>
